### ISSUE TITLE
Smart brackets are sometimes too smart

### ISSUE THREAD

**Issue by cpitclaudel** [greghendershott/racket-mode#210]
With this sample:

```
#lang racket

(regexp-match* #px"\\{|\\}|\\+|\\*|,| +|@[a-zA-Z+|[a-zA-Z+
```

Typing `]` at the end of the line inserts a `}`.

**Comment by cpitclaudel** [comment]
Same with this:

```
(eq (string-ref str #\
```

**Comment by greghendershott** [comment]
Thanks for the report.

- Possibly many people haven't experienced this because they use one of `paredit` (I do), `smartparens`, or even just `electric-pair-mode` (in Emacs 24.5+, I think).

-  There's also a `racket-smart-open-bracket` feature, similar to what DrRacket does for _open_ delimiters. But that feature is off by default, and even when enabled is N/A for this.

- `racket-mode` binds <kbd>)</kbd> <kbd>]</kbd> and <kbd>}</kbd> to the `racket-insert-closing` command, which is what's doing this.  (In all cases remember you can <kbd>C-q</kbd> or <kbd>C-u</kbd> then type the exact thing you want.)

  - The first problem example could be addressed by disabling the special behavior in strings.

  - The second example could be addressed by having it check for `#\` Racket character literals.

---

p.s.

> though shouldn't typing } just insert it in all cases?

I don't think so. Racket treats `}` as equivalent to `)` and `]`. You can use any of those for any s-expression, so long as they match.  There are some _conventions_ for when to use each, but they're just conventions in `#lang racket`.  Some `#lang`s treat it differently, but I think it's reasonable to assume it would always be a matching delimiter.

**Comment by cpitclaudel** [comment]
Thanks for the reply! Not sure about the matching delimiter part (in scribble in particular) but I sounds like a good approximation any way.
My question about '}' was more along the lines of: "what's the point of having two separate key do the same?" Presumably if I want smart parens I can press `)`, so wouldn't pressing `}` indicate that I don't?

**Comment by greghendershott** [comment]
> My question about '}' was more along the lines of: "what's the point of having two separate key do the same?" Presumably if I want smart parens I can press ), so wouldn't pressing } indicate that I don't?

Ah OK, I see what you mean. That's a great question. I don't have a great answer. :) This was one of the very first things I added to racket-mode. I had followed the example here of [Quack](http://www.neilvandyke.org/quack/), which I had been using with `scheme-mode`. It bound all 3 keys.

I can see <kbd>]</kbd> because it's un-shifted and fast. I can see <kbd>)</kbd> because people might type that out of habit. I guess the rationale is consistency -- since Scheme (and Racket) treat them all the same, may as well bind <kbd>}</kbd>, too?

Since people can rebind anything in `racket-mode-hook`, and since IMHO an even better idea is to use `paredit` or similar? I think I'll leave the default keybindings as-is, I'll push a commit simply handling the in-string and after-`#\` examples you provided.
